### PR TITLE
Fix 2 issues during recent upgrade

### PR DESCRIPTION
**Pull request by yizheliu-amazon**
*Issue #, if available:*

*Description of changes:*
Fix 2 issues during recent upgrade
- when CategoryField is checked, and then unchecked, in parent <formik> will still have validation error saying category field is empty:
  - root cause is `setFieldValue` will [trigger validation](https://formik.org/docs/api/formik#validateonchange-boolean) if `validationOnChange` to `true`(default value). Changing `validationOnChange` to `false` fix the issue
- backend failure is propagated as success to frontend
 - root cause is returning `kibanaResponse.ok` when error is found like [such line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/server/routes/ad.ts#L202-L206) will NOT reach to [FAILURE](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/reducers/ad.ts#L148) type in reducer. Instead, it will reach to SUCCESS type. For example, given user has no permission to stop/start detector, after such user starts detector, UX says detector is started, even though backend throws exception, and detector is NOT started.

Before fix
Screen Shot 2020-12-02 at 2 53 03 PM
After fix
Screen Shot 2020-12-02 at 2 55 05 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: public/pages/EditFeatures/containers/EditFeatures.tsx [hunk 1]
```diff
@@ -376,6 +382,7 @@ export function EditFeatures(props: EditFeaturesProps) {
           handleSubmit(values, actions.setSubmitting)
         }
         validate={validateFeatures}
+        validateOnChange={false}
```
Comment (yizheliu-amazon) [inline public/pages/EditFeatures/containers/EditFeatures.tsx:385]: This is the only change to fix formik validation issue for CategoryField

**Comment by yizheliu-amazon** [comment]
Have verified unit test and e2e test passed.

**Comment by yizheliu-amazon** [comment]
will add change for release note in next commit.

**Comment by ohltyler** [comment]
Regarding the issue of returning `kibanaResponse.ok` in the failure case, did you ever look into changing [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/middleware/clientMiddleware.ts#L44) and [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/middleware/clientMiddleware.ts#L49)?

Looks like I may have forgotten to remove the `data.` prefix there, which was changed with the new HTTP client. For example, I've tested locally and gotten it to properly go to the `FAILURE` reducer type by changing `data.ok` to `ok`

File: public/pages/EditFeatures/containers/EditFeatures.tsx (hunk 1 above)
Comment (ylwu-amzn) [reply public/pages/EditFeatures/containers/EditFeatures.tsx:385]: Does this `validateOnChange` only control validating CategoryField or not? Just make sure we don't ignore validation for other parts which necessary.

File: server/routes/utils/adHelpers.ts [hunk 2]
```diff
@@ -253,3 +254,55 @@ export const getErrorMessage = (err: any) => {
...
     : get(err, 'message');
 };
+
+export const getKibanaErrorResponseByStatusCode = (
+  statusCode: number,
+  message: string
+) => {
+  switch (statusCode) {
+    case 400:
+      return kibanaResponseFactory.badRequest({
```
Comment (ylwu-amzn) [inline server/routes/utils/adHelpers.ts:264]: Just curious, why need to explicitly return `kibanaResponseFactory.badRequest`, looks like something duplicate for `400` and `badRequest` Is there any general error response to accept status code ?

File: public/pages/EditFeatures/containers/EditFeatures.tsx (hunk 1 above)
Comment (yizheliu-amazon) [reply public/pages/EditFeatures/containers/EditFeatures.tsx:385]: Yes. For `Features`, its `validateOnChange` is set to be true in [code here](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/da5ead30a8266f43ff5212d59f18ec75f4559a8b/public/pages/EditFeatures/containers/EditFeatures.tsx#L155). And when save button is clicked, it will do check for category field again in [handleSaveChanges](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/da5ead30a8266f43ff5212d59f18ec75f4559a8b/public/pages/EditFeatures/containers/EditFeatures.tsx#L278).

**Comment by yizheliu-amazon** [comment]
> Regarding the issue of returning `kibanaResponse.ok` in the failure case, did you ever look into changing [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/middleware/clientMiddleware.ts#L44) and [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/middleware/clientMiddleware.ts#L49)?
> 
> Looks like I may have forgotten to remove the `data.` prefix there, which was changed with the new HTTP client. For example, I've tested locally and gotten it to properly go to the `FAILURE` reducer type by changing `data.ok` to `ok`, since it is now finding the `ok: false` field inside of the Kibana response.
> 
> I believe the same goes for changing `data.error` to `error`.

Good catch. Removing `data.` prefix works.

File: server/routes/utils/adHelpers.ts (hunk 2 above)
Comment (yizheliu-amazon) [reply server/routes/utils/adHelpers.ts:264]: `kibanaResponseFactory` has multiple response items with corresponding statusCode, `badRequest` has 400 status code. as shown like [this](https://github.com/elastic/kibana/blob/v7.10.0/src/core/server/http/router/response.ts#L155). This method is for given status code, return corresponding response.

But given Tyler's found above, I will revert this change. No need to worry about it.

**Comment by ohltyler** [comment]
> > Regarding the issue of returning `kibanaResponse.ok` in the failure case, did you ever look into changing [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/middleware/clientMiddleware.ts#L44) and [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/middleware/clientMiddleware.ts#L49)?
> > Looks like I may have forgotten to remove the `data.` prefix there, which was changed with the new HTTP client. For example, I've tested locally and gotten it to properly go to the `FAILURE` reducer type by changing `data.ok` to `ok`, since it is now finding the `ok: false` field inside of the Kibana response.
> > I believe the same goes for changing `data.error` to `error`.
> 
> Good catch. Removing `data.` prefix works.

Cool, so seems we can revert the Kibana response changes on server side, and the error parsing changes on the browser side then?

**Comment by yizheliu-amazon** [comment]
> > > Regarding the issue of returning `kibanaResponse.ok` in the failure case, did you ever look into changing [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/middleware/clientMiddleware.ts#L44) and [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/master/public/redux/middleware/clientMiddleware.ts#L49)?
> > > Looks like I may have forgotten to remove the `data.` prefix there, which was changed with the new HTTP client. For example, I've tested locally and gotten it to properly go to the `FAILURE` reducer type by changing `data.ok` to `ok`, since it is now finding the `ok: false` field inside of the Kibana response.
> > > I believe the same goes for changing `data.error` to `error`.
> > 
> > 
> > Good catch. Removing `data.` prefix works.
> 
> Cool, so seems we can revert the Kibana response changes on server side, and the error parsing changes on the browser side then? Apologize for missing this on the initial commits.

Right, I will revert those lines and make it way simpler.

No worries. I think it is not easy for find such issue, another data point for more automated test.

File: public/pages/EditFeatures/containers/EditFeatures.tsx (hunk 1 above)
Comment (yizheliu-amazon) [reply public/pages/EditFeatures/containers/EditFeatures.tsx]: sorry for reverting the change. has found it can not fix the issue completely. Has made such change to completely fix the issue: https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/pull/339/files#diff-b12f07c85b6d89dbdd055f0482e17c93168efdabb5b2d08ec6fde29ee7b6dfddR292

File: public/redux/middleware/__tests__/clientMiddleware.test.ts [hunk 3]
```diff
@@ -92,12 +92,12 @@ describe('client MiddleWare', () => {
...
     const actionCall = jest
       .fn()
-      .mockReturnValue(fooPromise(true, { ok: false }));
+      .mockReturnValue(fooPromise(true, { body: { ok: false } }));
     const action = { type: 'TEST', request: actionCall };
     return invoke(action).then(() => {
       expect(next.mock.calls.length).toBe(2);
       expect(next.mock.calls[1][0]).toEqual({
-        result: { ok: false },
+        result: { body: { ok: false } },
```
Comment (ohltyler) [inline public/redux/middleware/__tests__/clientMiddleware.test.ts:100]: Are these changes still necessary?

Comment (yizheliu-amazon) [reply public/redux/middleware/__tests__/clientMiddleware.test.ts:100]: yes. This is to fix the failed ut.

**Review by ohltyler — APPROVED** [review]
Thanks for the fix!

**Review by ylwu-amzn — APPROVED** [review]
